src/vcpu.rs

```diff
@@ -365,47 +369,92 @@ impl RISCVVCpu {
+                            // we do not set `a1` here, because guest VM may not expect a modification to `a1`.
+                        }
+                        legacy::LEGACY_CONSOLE_GETCHAR => {
+                            // sbi_call_legacy_0(LEGACY_CONSOLE_GETCHAR)
+                            let c: isize = -1;
+                            let ret = sbi_rt::console_read(sbi_rt::Physical::new(
+                                1,
+                                crate_interface::call_interface!(crate::HalIf::virt_to_phys(
+                                    HostVirtAddr::from_ptr_of(core::ptr::addr_of!(c))
+                                ))
+                                .as_usize(),
+                                0,
+                            ));
+                            if ret.is_ok() {
+                                self.set_gpr_from_gpr_index(GprIndex::A0, c as _);
+                            } else {
+                                warn!(
                                 "LEGACY_CONSOLE_GETCHAR c {:#x} param {:#x?} err {:#x} value {:#x}",
                                 c, param, ret.error, ret.value
                             );
-                            self.set_gpr_from_gpr_index(GprIndex::A0, ret.error);
+                                self.set_gpr_from_gpr_index(GprIndex::A0, ret.error);
+                            }
                         }
-                    }
-                    sbi_spec::legacy::LEGACY_SHUTDOWN => {
-                        // sbi_call_legacy_0(LEGACY_SHUTDOWN)
-                        return Ok(AxVCpuExitReason::SystemDown);
+                        legacy::LEGACY_SHUTDOWN => {
+                            // sbi_call_legacy_0(LEGACY_SHUTDOWN)
+                            return Ok(AxVCpuExitReason::SystemDown);
+                        }
+                        _ => {
+                            warn!(
+                                "Unsupported SBI legacy extension id {:#x} function id {:#x}",
+                                extension_id, function_id
+                            );
+                        }
+                    },
+                    // Handle HSM extension
+                    hsm::EID_HSM => match function_id {
+                        hsm::HART_START => {
+                            let hartid = a[0];
+                            let start_addr = a[1];
+                            let opaque = a[2];
+                            self.advance_pc(4);
+                            return Ok(AxVCpuExitReason::CpuUp {
+                                target_cpu: hartid as _,
+                                entry_point: GuestPhysAddr::from(start_addr),
+                                arg: opaque as _,
+                            });
+                        }
+                        hsm::HART_STOP => {
+                            return Ok(AxVCpuExitReason::CpuDown { _state: 0 });
```

Comment:
Do we need `self.advance_pc(4);` here?

Reply:
Nope, the `HART_STOP`'s semantic is totally different from `HART_START`.

`HART_START` will bring up a secondnary core which is not current one, the current core still need to resume running, cause `self.advance_pc(4)` is needed

 `HART_STOP` will stop current core (through vcpu wait & destory), without resume the context of the current core, if current core is brought back to running, it's context has to be reset, so `self.advance_pc(4)` is not needed here